src/magic_folder/test/test_snapshot.py

```diff
@@ -0,0 +1,118 @@
+import io
+import os
+
+from twisted.internet import defer
+from twisted.python.filepath import (
+    FilePath,
+)
+from twisted.web.resource import (
+    Resource,
+)
+from twisted.web.client import (
+    Agent,
+    FileBodyProducer,
+)
+
+from treq.client import (
+    HTTPClient,
+)
+from treq.testing import (
+    RequestTraversalAgent,
+    RequestSequence,
+    StubTreq,
+)
+
+from testtools.matchers import (
+    StartsWith,
+)
+
+from magic_folder.snapshot import create_snapshot
+
+from .fixtures import (
+    NodeDirectory,
+)
+from .common import (
+    ShouldFailMixin,
+    SyncTestCase,
+    AsyncTestCase,
+    skipIf,
+)
+
+
+class _FakeTahoeRoot(Resource):
+    """
+    This is a sketch of how an in-memory 'fake' of a Tahoe
+    WebUI. Ultimately, this will live in Tahoe
+    """
+
+
+class _FakeTahoeUriHandler(Resource):
+    """
+    """
+
+    isLeaf = True
+
+    def render(self, request):
+        return b"URI:DIR2-CHK:some capability"
+
+
+def create_fake_tahoe_root():
+    """
+    Probably should take some params to control what this fake does:
+    return errors, pre-populate capabilities, ...
+    """
+    root = _FakeTahoeRoot()
+    root.putChild(
+        b"uri",
+        _FakeTahoeUriHandler(),
+    )
+    return root
```

Comment:
Your last suggestion is maybe-kind-of like the treq's `RequestSequence`..?

Looking at this all from magic-folder's perspective, it doesn't care about "the state of the grid" it just cares whether the next API request behaves like it expects. What I mean is, if I add a bunch of snapshots to the grid, I don't care what happens as long as a `"GET <some cap>"` returns the data I expect.

So what I was *hoping* would be a path forward would be a very minimal implementation of the first idea: we implement just `/uri` and just the two bits we need (POST with `?t=mkdir-immutable` and POST with `?t=upload`) and the "confirmation" side which is just GET on `/uri/<some capability>`. I believe that's a complete list of the WebUI methods that magic-folder will need to interact with.

I'd love to see an all-JSON API for Tahoe (*with* testing fakes) but I feel like that's probably "too much" work for immediate use in magic-folder? (Especially with, hopefully, ongoing py3-porting happening).

Reply:
> Looking at this all from magic-folder's perspective, it doesn't care about "the state of the grid" it just cares whether the next API request behaves like it expects. What I mean is, if I add a bunch of snapshots to the grid, I don't care what happens as long as a "GET <some cap>" returns the data I expect.

That's true as far as it goes.  However, how do we know if Tahoe-LAFS will return the data magic-folder expects for a `GET <some cap>` request?  I think the only way to know is to simulate a non-trivial amount of Tahoe-LAFS behavior.

`GET <some cap>` will only return what magic-folder expects if magic-folder previously did the right kind of `PUT ...`, right?

The mock style of testing would be to pre-define the PUT that we think is required and have it hand back the canned response we think is right.  I suppose you could verify this mock against the real Tahoe-LAFS but it's such a limited value mock (where could you use it except in one or two magic-folder tests?) that the cost of verification might be prohibitive.

It might be better to try to discuss this more in real time.

Reply:
> `GET <some cap>` will only return what magic-folder expects if magic-folder previously did the right kind of `PUT ...`, right?

What we probably need is some sort of a state machine driven Tahoe-LAFS side mocking that moves in lock-step with the client's state machine perhaps?

Reply:
> GET <some cap> will only return what magic-folder expects if magic-folder previously did the right kind of PUT ..., right?

Yes.

I mean, the functionality that magic-folder depends on here IS pretty simple, really. It also would probably be good to discuss real-time though. But, yeah, the "meat" here is to simply hide all the complexity of Tahoe behind the couple APIs that magic-folder actually uses .. and of course some integration-tests (in tahoe) that show those API fakes do in fact do the same thing as a "real" grid does.

I think this would still let us do the "interesting" tests in magic-folder (like "alice changes stuff, then bob changes stuff before seeing alice's changes" etc) because we control when magic-folder updates.

Reply:
Okay, here's some more meat (mostly around "create fake data but with real capability-strings"). I think we only need 3 APIs to work for magic-folder: https://github.com/tahoe-lafs/tahoe-lafs/pull/723